libfaketime.8
.\" libfaketime.8 --
.\"
.\" Copyright (C) 2002-2004,2023 Krzysztof Burghardt <krzysztof@burghardt.pl>.
.\"
.TH LIBFAKETIME 8 "19 Feb 2023" "Linux`s Utilities" "Linux Users` Manual"
.SH NAME
libfaketime \- a library that adds a filter to time() function calls
.SH SYNOPSIS
.BI "it`s a library so you cannot execute it directly"
.SH DESCRIPTION
.B libfaketime
is a
.BR SO\ library
that provides filters in the
.BR time()
function.
.SS
libfaketime features:
.HP
- returns real or fake time to programs,
.HP
- has user-wide (~/.libfaketime.d/) and system-wide (/etc/libfaketime.d/) configuration,
.HP
- works transparently for programs,
.SH OPTIONS
Library does not need any options.
.SH Where\ to\ get\ libfaketime
.PP
The source code for libfaketime can be found at
https://github.com/burghardt/libfaketime
.SH Installing\ libfaketime
.TP 3
.TP
1.
If you install libfaketime on Linux, ensure that your shared loader
(ld-linux.so.1/ld.so) supports \fR\&\f(CWLD_PRELOAD\fR.
Recommended ld.so version 1.8.5 or never.
.TP
2.
Change cwd to the source code directory.
.TP
3.
Type \fR\&\f(CWmake\fR to compile and build libfaketime.so.1.
.TP
4.
Type \fR\&\f(CWmake install\fR to install the libfaketime library and manual
in their final destination locations.
.TP
5.
To use libfaketime, set the environment variable \fR\&\f(CWLD_PRELOAD\fR to
point to the libfaketime library. Example (sh syntax):
.IP

.nf
.ft 3
.in +0.3i
      LD_PRELOAD=libfaketime.so.1
      export LD_PRELOAD
.fi
.in -0.3i
.ft R
.lp

\&\fR
.IP
or (csh syntax):
.IP

.nf
.ft 3
.in +0.3i
      setenv LD_PRELOAD libfaketime.so.1
.fi
.in -0.3i
.ft R
.lp

.IP
You might want to put these lines in your \fR\&\f(CW.profile\fR or
\&\fR\&\f(CW.cshrc\fR in order to activate libfaketime for all processes that you
initiate.
.TP
6.
Use your programs as you would normally. The libfaketime will transparently check the
time() callers, and if detected that should report fake time will change the cloc value for this program.
If a filtered time() is used, the following will occur:
.RS
.PD 0
.TP 3
1.
If the library detects that fake time should be returned and syslog support was enabled an entry will be added to /var/log/messages.
The following is an example of such an entry:
.RS
Mar 26 20:20:22 techie Libfaketime: using FAKE time() for (date:14654) UID(500) EUID(0)
.RE
.TP
2.
If the library cannot get fake time, and -DKILLER was defined in compile flags library will call _exit().
(To prevent making timestamps.)
.TP
.RE
.BR
.PP
For security reasons, the dynamic loader disregards environmental
variables such as \fR\&\f(CWLD_PRELOAD\fR when executing set-uid programs.
However, on Linux, you can use libfaketime with set-uid programs too, by using
one of the two methods described below:
.TP 3
1.
You may append the path to \fR\&libfaketime.so.1\fR into
\&\fR\&\f(CW/etc/ld.so.preload\fR instead of using
\fR\&\f(CWLD_PRELOAD\fR.
.IP
\&\fBWARNING\fR: If you use \fR\&\f(CW/etc/ld.so.preload\fR, be sure to
install \fR\&\f(CWlibfaketime.so.1\fR on your root filesystem, for
instance in \&\fR\&\f(CW/lib\fR, as is done by the default installation. Using
a directory which is not available at boot time, such as /usr/local/lib will
cause trouble at the next reboot! You also must not define -DKILLER if you want
to use libfaketime loaded from \fR\&\f(CW/etc/ld.so.preload\fR, because it will
attempt to kill your /sbin/init process when booting.
.IP
You should also be careful to remove libfaketime from
\fR\&\f(CW/etc/ld.so.preload\fR when installing a new version.  First
test it out using \fR\&\f(CWLD_PRELOAD\fR, and only if everything is OK, put it
back into \&\fR\&\f(CW/etc/ld.so.preload\fR.

.TP
2.
If you have a version of \fR\&\f(CWld.so\fR which is more recent than
\&\fR\&\f(CW1.9.0\fR, you can set \fR\&\f(CWLD_PRELOAD\fR to contain the
basename of \fR\&\f(CWlibfaketime.so.1\fR without the directory.  In that
case, the file is found as long as it is in the shared library path
(usually containing \fR\&\f(CW/lib\fR and \fR\&\f(CW/usr/lib\fR)).
Because the search is restricted to the library search path, this also works
for set-uid programs.
Example (sh syntax):
.IP

.nf
.ft 3
.in +0.3i
      LD_PRELOAD=libfaketime.so.1
      export LD_PRELOAD
.fi
.in -0.3i
.ft R
.lp

\&\fR
.IP
or (csh syntax):
.IP

.nf
.ft 3
.in +0.3i
      setenv LD_PRELOAD libfaketime.so.1
.fi
.in -0.3i
.ft R
.lp

\&\fR
.IP
The advantage of this approach over \fR\&\f(CWld.so.preload\fR is that
libfaketime can more easily be switched off in case something goes wrong.
On the other hand, LD_PRELOAD can be easily unset by programs, and time()
 hooking will not work anymore.
.PP

.SH Using libfaketime
.iX "c Symbolic links"
.PP
Once libfaketime is installed and either LD_PRELOAD or /etc/ld.so.preload has been
appropriately configured (e.g. in compile time you can define two compilation flags),
there is something else to do.  The processes to be monitored must have their files
in the configuration directory (e.g. ~/.libfaketime.d/ for user-wide and
/etc/libfaketime.d/ for system-wide configuration).
.PP
Now you can put <programname>.nofake to the system configuration directory,
which prohibits users from faking time for this program (useful for mailers,
which should have a valid date and time).
.PP
In the user and system-wide configuration directory, you can put <programname>.fake
with time in seconds since 00:00:00, Jan 1, 1970. How to get a timestamp? Run
.BR date\ -d\ "31\ July\ 1983\ 5:40"\ +%s

.PP
.SH How\ it\ works
libfaketime uses a novel method for detecting and handling time(). First, this
technique intercepts all calls to library functions known to return time/date.
Then, a substituted version of the corresponding function implements the
original functionality but in a manner that ensures that only the program
not listed in *.fake files can get a valid (actual) time/date. Read sources
for more information.
.SH FILES
.HP
.I /{lib,lib64}/libfaketime.so.<version>
- main library
.HP
.I /usr/man/man8/libfaketime.8
- manual page
.HP
.I /etc/libfaketime.d/
- system-wide configuration directory,
.HP
.I ~/.libfaketime.d/
- user-wide configuration directory,
.SH "FILE FORMAT"
.I ~/.libfaketime.d/
and
.I /etc/libfaketime.d/
.P
You have to configure a separate file for each program.
.P
For example:
.P
write into the file named
.I date.fake
fixed time() value which date will always get.
.SH "SEE ALSO"
.BR libsafe (8),
.BR date (1),
.BR time (2),
.BR dlsym (3),
.BR ld.so (8),
.BR ldd (1)
.SH HISTORY
The 1st libfaketime version (1.0.0) appeared in 20 May 2002.
.SH BUGS
Report bugs using GitHub's project issues section.
.SH AUTHOR
This library was written by Krzysztof Burghardt <krzysztof@burghardt.pl>
and may be freely distributed under the terms of the GNU General Public License.
There is ABSOLUTELY NO WARRANTY for this program.


ChangeLog
2023-02-19 Krzysztof Burghardt
	* libfaketime.c: allow ld.so to search for libc.so.6
	  (/lib absolute path removed to support multi-arch)

2007-11-01 Krzysztof Burghardt
	* libfaketime.c: add support for clock_gettime()

2004-09-21 Krzysztof Burghardt
	* libfaketime.c: fixed to be "thread safe"

2004-03-25 Krzysztof Burghardt
	* libfaketime.c: added support for gettimeofday with
	  working tv_usec. (Without it _most_ programs failed to work.)

2003-03-26 Krzysztof Burghardt
	* libfaketime.c: fixed (annother) BUG in get_exe code when
	  running SUID/SGID binaries on new kernels (e.g. 2.2.26pre1 and
	  2.4.21pre) or any kernels witch Alan patch for ptrace vulnerability
	  (e.g. CAN-2003-0127),

2002-05-21 Krzysztof Burghardt
	* libfaketime.c: fixed BUG in get_exe code when running
	  SUID/SGID binaries,
	* libfaketime.c: fixed BUG in time() now it changing value
	  no delta detection no time freezes,
	* libfaketime.c: caching time readed from file
	  (only one file access per one proces, in older it was one file access
	   per one time() call which is not a best idea (programs works slow
	   and HDD was use too much)),
	* libfaketime.c: added some debug code,
	* libfaketime.8: added man page :-)

2002-05-20 Krzysztof Burghardt
	* src/lib/libfaketime.c: loading from /etc/ld.so.preload file
	  (against SUID/SGID binaries),


libfaketime.c
/*
   LibFakeTime library - fake time() library
   Copyright (C) 2002-2004,2007,2023 Krzysztof Burghardt <krzysztof@burghardt.pl>

   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "libfaketime.h"

static void _libfaketime_init (void) __attribute__((constructor));
static void _libfaketime_fini (void) __attribute__((destructor));

extern char **environ[];

static int init_done = FALSE;	/* return 1 if _init was run, or 0 otherwise */
static int real_time = TRUE;	/* return real time if 1, or fake if 0 */
static time_t fake_time = 0;	/* fake time readed from file */
void *lib_handle;		/* system library handle */
time_t (*real_time_handle) (time_t *) = NULL;	/* system time() handle */
int (*real_gettimeofday_handle) (struct timeval * tv, struct timezone * tz) = NULL;	/* system gettimeofday() handle */

static void *
xmalloc (int n)
{
  void *p;

  p = (void *) malloc (n);

  if (!p)
    {
      DEBUG_MESSAGE
	("LibFakeTime: xmalloc(): cannot alocate %i bytes...\n", n);
#ifdef KILLER
      /*
       * Should it kill proces?!
       * If proces have SIGSEGV (11) handled it can persist exeption...
       * (if not - glibc will kill it...)
       */
      DEBUG_MESSAGE ("LibFakeTime: xmalloc(): calling _exit()...\n");
      _exit (EXIT_FAILURE);
#endif
    }

  memset (p, 0, n);

  return p;
}

static char *
mk_path (char *path, const char *dir, char *file, const char *surfix)
{
  sprintf (path, "%s/%s.%s", dir, basename (file), surfix);

  DEBUG_MESSAGE
    ("LibFakeTime: mk_path(): dir=%s, file=%s, surfix=%s\n",
     dir, file, surfix);
  DEBUG_MESSAGE ("LibFakeTime: mk_path(): returning %s\n", path);

  return path;
}

static char *
get_exe (char *exe, int size)
{
  char *temp;
  temp = xmalloc (size);

  if (readlink ("/proc/self/exe", temp, size) <= 0)
    {
      exe[0] = '\0';
      DEBUG_MESSAGE ("LibFakeTime: get_exe(): cannot get exe name...\n");
#ifdef KILLER
      /*
       * It`s impossible?!
       * If you have /proc/ NOT mounted it can kill your proces...
       * (e.g. init while booting and proc is not mounted yet)
       */
      DEBUG_MESSAGE ("LibFakeTime: get_exe(): calling _exit()...\n");
      free (temp);
      temp = NULL;
      _exit (EXIT_FAILURE);
#endif
    }
  else
    {
      sprintf (exe, "%s", basename (temp));
      DEBUG_MESSAGE ("LibFakeTime: get_exe(): returning %s\n", exe);
    }
  free (temp);
  temp = NULL;

  return exe;
}

static int
can_i_fake_test (const char *dir, char *exe)
{
  FILE *fd;
  char *path;
  int ret = FALSE;

  path = xmalloc (1024);

  fd = fopen (mk_path (path, dir, exe, NO_FAKE_SURFIX), "r");
  if (!fd)
    {
      DEBUG_MESSAGE
	("LibFakeTime: can_i_fake_test(): cannot open %s\n", path);
      DEBUG_MESSAGE ("LibFakeTime: can_i_fake_test(): returning TRUE\n");
      ret = TRUE;
    }
  else
    {
      DEBUG_MESSAGE ("LibFakeTime: can_i_fake_test(): opened %s\n", path);
      DEBUG_MESSAGE ("LibFakeTime: can_i_fake_test(): returning FALSE\n");
      fclose (fd);
    }
  free (path);
  path = NULL;

  return ret;
}

static time_t
get_fake_time (const char *dir, char *exe)
{
  FILE *fd;
  time_t read_fake_time = 0;
  static const int buf_len = 1024;
  char *buf;
  char *path;

  buf = xmalloc (buf_len);
  path = xmalloc (1024);

  fd = fopen (mk_path (path, dir, exe, FAKE_SURFIX), "r");
  if (!fd)
    {
      read_fake_time = 0;
      DEBUG_MESSAGE ("LibFakeTime: get_fake_time(): cannot open %s\n", path);
    }
  else
    {
      if (fgets (buf, buf_len - 1, fd) != NULL)
	{
	  buf[strlen (buf) - 1] = '\0';
	  read_fake_time = (time_t) atoi (buf);
	}
      else
	{
	  read_fake_time = 0;
	  DEBUG_MESSAGE ("LibFakeTime: get_fake_time(): cannot read %s\n",
			 path);
	}
      fclose (fd);
      DEBUG_MESSAGE ("LibFakeTime: get_fake_time(): opened %s\n", path);
      DEBUG_MESSAGE ("LibFakeTime: get_fake_time(): readed %s\n", buf);
    }
  free (buf);
  buf = NULL;
  free (path);
  path = NULL;

  DEBUG_MESSAGE
    ("LibFakeTime: get_fake_time(): returning %i\n",
     (unsigned int) read_fake_time);
  return read_fake_time;
}

time_t
time (time_t * t)
{
  static time_t time_to_return = 0;

  /* Check if we are  already after _init and do _init if not yet */
  if (!init_done)
    _libfaketime_init ();

  if (real_time)
    {
      time_to_return = (*real_time_handle) (NULL);
      DEBUG_MESSAGE
	("LibFakeTime: time(): returning REAL time (%i)\n",
	 (unsigned int) time_to_return);
    }
  else
    {
      time_to_return = (*real_time_handle) (NULL) - fake_time;
      DEBUG_MESSAGE
	("LibFakeTime: time(): returning FAKE time (%i)\n",
	 (unsigned int) time_to_return);
    }

  if (t != 0)
    {
      DEBUG_MESSAGE ("LibFakeTime: time(): setting *t...\n");
      *t = time_to_return;
    }
  return time_to_return;
}

int
gettimeofday (struct timeval *tv, void *tz)
{
  static int retval = 0;

  /* Check if we are  already after _init and do _init if not yet */
  if (!init_done)
    _libfaketime_init ();

  retval = (*real_gettimeofday_handle) (tv, tz);
  DEBUG_MESSAGE ("LibFakeTime: gettimeofday(): retval = %i...\n", retval);
  if (!retval)
    {
      /* do not touch tv_usec - it`s extremaly important to return
         _VALID_ tv_usec due to many programs uses this for hi-res timing
         only tv_sec can be (safely) affected */
      DEBUG_MESSAGE ("LibFakeTime: gettimeofday(): setting *tv->tv_sec...\n");
      tv->tv_sec = time (NULL);
    }
  DEBUG_MESSAGE ("LibFakeTime: gettimeofday(): returning retval...\n");
  return retval;
}

int
clock_gettime (clockid_t clk_id __attribute__((unused)), struct timespec *tp)
{
  static int retval = 0;

  /* Check if we are  already after _init and do _init if not yet */
  if (!init_done)
    _libfaketime_init ();

  /* Just to be sure... */
  if (tp == NULL)
    {
      DEBUG_MESSAGE
	("LibFakeTime: clock_gettime(): returning due to NULL pointers...\n");
      return 0;
    }
  struct timeval tv;
  retval = gettimeofday (&tv, NULL);
  DEBUG_MESSAGE ("LibFakeTime: clock_gettime(): retval = %i...\n", retval);
  if (!retval)
    {
      /* do not touch tv_usec - it`s extremaly important to return
         _VALID_ tv_usec due to many programs uses this for hi-res timing
         only tv_sec can be (safely) affected */
      DEBUG_MESSAGE
	("LibFakeTime: clock_gettime(): setting *tp->tv_(u)sec...\n");
      tp->tv_sec = tv.tv_sec;
      tp->tv_nsec = tv.tv_usec * 1000;
    }
  DEBUG_MESSAGE ("LibFakeTime: clock_gettime(): returning retval...\n");
  return retval;
};

static void
_libfaketime_init (void)
{
  static const int exe_len = 1024;
  time_t time_to_return = 0;
  char *error;
  char *exe;

  DEBUG_MESSAGE
    ("LibFakeTime: _init(): initalizing library (%i)...\n", VERSION);

  exe = xmalloc (exe_len);

  lib_handle = dlopen ("libc.so.6", RTLD_LAZY);
  if (!lib_handle)
    {
      fprintf (stderr, "%s\n", dlerror ());
      exit (1);
    }

  real_time_handle = dlsym (lib_handle, "time");
  if ((error = dlerror ()) != NULL)
    {
      fprintf (stderr, "%s\n", error);
      exit (1);
    }

  real_gettimeofday_handle = dlsym (lib_handle, "gettimeofday");
  if ((error = dlerror ()) != NULL)
    {
      fprintf (stderr, "%s\n", error);
      exit (1);
    }

  get_exe (exe, exe_len);

  if ((real_time = can_i_fake_test (SYS_FAKE_TIME_DIR, exe)))
    {
      char *user_path;
      user_path = xmalloc (1024);

      sprintf (user_path, "%s/%s", getenv ("HOME"), USER_FAKE_TIME_DIR);
      time_to_return = get_fake_time (user_path, exe);
      free (user_path);
      user_path = NULL;

      if (!time_to_return)
	{
	  time_to_return = get_fake_time (SYS_FAKE_TIME_DIR, exe);
	}
    }

  if (time_to_return)
    {
      real_time = FALSE;
#ifdef HAVE_SYSLOG_H
      openlog ("LibFakeTime", LOG_CONS | LOG_NDELAY, LOG_USER);
      syslog (LOG_WARNING,
	      "using FAKE time() for (%s:%i) UID(%i) EUID(%i)",
	      exe, getpid (), getuid (), geteuid ());
      closelog ();
#endif
      fake_time = (*real_time_handle) (NULL) - time_to_return;
    }
  else
    {
      real_time = TRUE;
    }
  free (exe);
  exe = NULL;

  init_done = TRUE;
}

static void
_libfaketime_fini (void)
{
  DEBUG_MESSAGE ("LibFakeTime: _fini(): finalizing library...\n");
  dlclose (lib_handle);
}
